Fix typed array export

ajreckof opened the pull request:
Fixes my mess from #76114
Fixes #75981 (as a side effect, if this PR is not accepted i could easily extract the fix from it)
Fixes also the bug i found and reported here : https://github.com/godotengine/godot/pull/76378#issuecomment-1519184710
Edit : those two bugs are related to the one just above and should be fixed with this PR #74141 #74050

This PR remove the hack with pointer and replace it with a proper handling from `EditorPropertyNodePath` of both `NodePath` and `Node`. Also needed to revise a bit the saving and loading of scene to handle properly the change.

PS: This is not polished and still needs testing

*Bugsquad edit:* Fixes #76642 Fixes #74141 Fixes #74050

Chaosus commented:
Please amend the commit name - no need to excuse in the title, while you can do that in the PR description. 😸

KoBeWi commented:
Needs rebase.

KoBeWi commented:
While testing another PR, I accidentally discovered that this PR breaks ViewportTexture. When you re-assign the viewport, the path will be relative to the node with texture instead of root node.

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -3701,6 +3677,36 @@ void EditorPropertyNodePath::_notification(int p_what) {
+		base_node = Object::cast_to<Node>(InspectorDock::get_inspector_singleton()->get_edited_object());
+	}
+	if (!base_node) {
+		//try a base node within history
```
```suggestion
		// Try a base node within history.
```

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -3701,6 +3677,36 @@ void EditorPropertyNodePath::_notification(int p_what) {
 }
+Node *EditorPropertyNodePath::get_base_node() {
+	Node *base_node = nullptr;
+	if (base_hint != NodePath()) {
+		if (get_tree()->get_root()->has_node(base_hint)) {
```
Can be merged into 1 condition.
```suggestion
	if (!base_hint.is_empty() && get_tree()->get_root()->has_node(base_hint)) {
```

ajreckof committed:
Fix typed array export

Apply suggestions from code review to squash later

Revert "Fix typed array export... again"

This reverts commit da8d6734fbc31f68e7e822f37fd239a92ac79b34.

Co-Authored-By: Tomek <kobewi4e@gmail.com>

KoBeWi reviewed (approved):
I tested and it fixes many issues with exported nodes and arrays.

The PR changes how exported nodes work (not only arrays), so needs a check from @reduz who implemented it originally.

ajreckof commented on editor/editor_properties.h:
```diff
@@ -781,20 +781,19 @@ class EditorPropertyNodePath : public EditorProperty {
 
-	String _get_meta_pointer_property() const;
 	virtual Variant _get_cache_value(const StringName &p_prop, bool &r_valid) const override;
-	virtual StringName _get_revert_property() const override;
```
Just realised while working on something else that what I was doing previously of returning `get_edited_property` is in fact the non-overriden behaviour so I just removed the whole overriding. Moreover I think this was the only place where this function was overriden so I'm wondering wether we should remove it to use `get_edited_property` everywhere it is used instead?

ajreckof replied:
I'm also wondering if we could remove the overriding of `_get_cache_value` but I'm not sure at 100% why there is a check for `p_prop == get_edited_property` but I think it was to update the cache when the property was being changed between pointer mode or not.

YuriSizov commented on editor/editor_properties.cpp:
```diff
@@ -3701,6 +3673,35 @@ void EditorPropertyNodePath::_notification(int p_what) {
 
 void EditorPropertyNodePath::_bind_methods() {
 }
+Node *EditorPropertyNodePath::get_base_node() {
```
Missing a new line, but also this method should be somewhere above, among other similar methods.

akien-mga replied:
Did the change so we can merge this now.

YuriSizov replied:
I don't see any changes 👀

akien-mga replied:
Welp, seems like I forgot to force push?

YuriSizov reviewed (approved):
I would like @reduz to take a look, but with editor part tested and core changes looking fine I think it's okay to merge for now and get him to confirm changes later. There aren't any public-facing changes or incompatibilities with existing scenes, so we can even revert this PR if there are major issues.

My note still needs to be addressed though.

akien-mga commented:
Thanks!

ajreckof commented:
If this is cherry picked. #78300 should be cherry-picked at the same time.

YuriSizov commented:
Actually, no, this won't be cherry-picked. See https://github.com/godotengine/godot/pull/76378#issuecomment-1594812425.
